hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -156,26 +152,50 @@ private ExecutionPlanBuilder() {
      */
     public static Map<MemberInfo, int[]> getPartitionAssignment(NodeEngine nodeEngine, List<MemberInfo> memberList) {
         IPartitionService partitionService = nodeEngine.getPartitionService();
-        Map<Address, MemberInfo> membersByAddress = memberList.stream().collect(toMap(MemberInfo::getAddress, identity()));
+        Map<Address, MemberInfo> membersByAddress = new HashMap<>();
+        for (MemberInfo memberInfo : memberList) {
+            membersByAddress.put(memberInfo.getAddress(), memberInfo);
+        }
 
-        final MemberInfo[] partitionOwners = new MemberInfo[partitionService.getPartitionCount()];
+        Map<MemberInfo, IntArrayWithPosition> partitionsForMember = new HashMap<>();
+        int partitionCount = partitionService.getPartitionCount();
         int memberIndex = 0;
-        for (int partitionId = 0; partitionId < partitionOwners.length; partitionId++) {
+
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
             Address address = partitionService.getPartitionOwnerOrWait(partitionId);
             MemberInfo member = membersByAddress.get(address);
             if (member == null) {
                 // if the partition owner isn't in the current memberList, assign to one of the other members in
                 // round-robin fashion
                 member = memberList.get(memberIndex++ % memberList.size());
             }
-            partitionOwners[partitionId] = member;
+            partitionsForMember.computeIfAbsent(member, ignored -> new IntArrayWithPosition(partitionCount)).add(partitionId);
+        }
+
+        Map<MemberInfo, int[]> partitionAssignment = new HashMap<>();
+        for (Entry<MemberInfo, IntArrayWithPosition> memberWithPartitions : partitionsForMember.entrySet()) {
+            partitionAssignment.put(memberWithPartitions.getKey(), memberWithPartitions.getValue().getFilledElements());
+        }
+        return partitionAssignment;
+    }
+
+    static class IntArrayWithPosition {
```

Comment:
I would use standard terminology here:
`IntArrayWithPosition` -> `FixedCapacityIntArrayList`
`size` -> `capacity`
`pos` -> `size`
`array` -> `elements`
`add(int i)` -> `add(int element)`
`getFilledElements` -> `asArray` (not using `toArray` as it doesn't always clone the internal array)

Also the `asArray` could assing `null` to `array` for better robustness - the list will be unusable after the call.